File: lib/lebai_tutorials/move_example.py
#!/usr/bin/env python2

import lebai_msgs
import rospy
from std_msgs.msg import String
from std_msgs.msg import Float64
from lebai_msgs.msg import ExecuteMode
from lebai_msgs.msg import MoveCommon
from lebai_msgs.msg import UntilInfo
from lebai_msgs.msg import IOConditionalExpress
from lebai_msgs.srv import MoveJoint
from lebai_msgs.srv import MoveLine
from lebai_msgs.srv import MoveCircle
from geometry_msgs.msg import Pose
from geometry_msgs.msg import Point
from geometry_msgs.msg import Quaternion


def SendMoveJoint():
    rospy.wait_for_service("/motion_service/move_joint")
    try:
        move_joint_caller = rospy.ServiceProxy('/motion_service/move_joint', MoveJoint)
        move_common = MoveCommon()
        move_common.is_continous_mode = False
        move_common.vel = 0.2
        move_common.acc = 0.5
        move_common.time = 0.0
        move_common.until = True
        move_common.until_info.io_express_logic = UntilInfo.LOGIC_AND
        express_1= IOConditionalExpress()
        express_1.group = 0
        express_1.type = IOConditionalExpress.TYPE_DIGITAL
        express_1.pin = 4
        express_1.uint_value = 1
        express_1.logic_operation = 2
        
        express_2 = IOConditionalExpress()
        express_2.group = 1
        express_2.type = IOConditionalExpress.TYPE_ANALOG
        express_2.pin = 1
        express_2.float_value= 3.579
        express_2.logic_operation = 3

        move_common.until_info.io_express = [express_1, express_2]
        move_joint_execute_mode = ExecuteMode()
        move_joint_execute_mode.data = ExecuteMode.EXECUTE_MODE_BUFFER

        # Call  the service
        result = move_joint_caller(is_joint_pose = True, joint_pose = [0.1, 0.2, 0.3, 0.4, 0.0, 0.0], 
        common = move_common, execute_mode = move_joint_execute_mode)

        rospy.loginfo("move result %d\n", result.ret)
        return
    except rospy.ServiceException as e:
        print("Service call failed: %s"%e)


def SendMoveLine():
    rospy.wait_for_service("/motion_service/move_line")
    try:
        move_line_caller = rospy.ServiceProxy('/motion_service/move_line', MoveLine)
        move_common = MoveCommon()
        move_common.is_continous_mode = False
        move_common.vel = 0.1
        move_common.acc = 1.0
        move_common.time = 0.0
        move_common.until = False
        move_line_execute_mode = ExecuteMode()
        move_line_execute_mode.data = ExecuteMode.EXECUTE_MODE_BUFFER

        move_line_cartesian_pose = Pose(position=Point(0.1,0.2,0.3), orientation=Quaternion(0.0, 0.0, 0.0, 1.0))
        # Call  the service
        result = move_line_caller(is_joint_pose = False, cartesian_pose = move_line_cartesian_pose, 
        common = move_common, execute_mode = move_line_execute_mode)

        rospy.loginfo("move result %d\n", result.ret)
        return
    except rospy.ServiceException as e:
        print("Service call failed: %s"%e)

def SendMoveCircle():
    rospy.wait_for_service("/motion_service/move_circle")
    try:
        move_circle_caller = rospy.ServiceProxy('/motion_service/move_circle', MoveCircle)
        move_common = MoveCommon()
        move_common.is_continous_mode = False
        move_common.vel = 0.1
        move_common.acc = 1.0
        move_common.time = 0.0
        move_common.until = False
        move_circle_execute_mode = ExecuteMode()
        move_circle_execute_mode.data = ExecuteMode.EXECUTE_MODE_BUFFER
        # Call  the service
        result = move_circle_caller(way_point_is_joint_pose = True, way_point_joint_pose=[0.1, 0.1, 0.1, 0.1,0.1, 0.1], 
        end_point_is_joint_pose = True, 
        end_point_joint_pose = [0.2, 0.2, 0.2, 0.2, 0.2, 0.2], 
        circle_angle = 0.3,
        common = move_common, execute_mode = move_circle_execute_mode)

        rospy.loginfo("move result %d\n", result.ret)
        return
    except rospy.ServiceException as e:
        print("Service call failed: %s"%e)
    
def Run():
    SendMoveJoint()
    SendMoveLine()
    SendMoveCircle()
    return

if __name__ == '__main__':
    try:
        rospy.init_node('move_example', anonymous=True)
        Run()
    except rospy.ROSInterruptException:
        pass

File: lebai_tutorials/src/move_example.cpp
#include <ros/ros.h>
#include <lebai_msgs/MoveJoint.h>
#include <lebai_msgs/MoveLine.h>

int main(int argc, char **argv)
{
    ros::init(argc, argv, "lebai_lm3_move_example");
    ros::NodeHandle n;

    ros::ServiceClient move_joint_client = n.serviceClient<lebai_msgs::MoveJoint>("/motion_service/move_joint");
    ros::ServiceClient move_line_client = n.serviceClient<lebai_msgs::MoveJoint>("/motion_service/move_line");
    move_joint_client.waitForExistence();
    ROS_INFO("Move joint service exist.\n");
    move_line_client.waitForExistence();
    ROS_INFO("Move line service exist.\n");

    //generate a move joint srv data.
    lebai_msgs::MoveJoint move_joint_srv;
    move_joint_srv.request.joint_pose = {-0.515705166127304, -1.383938292070577, 0.9317015810420413, -1.0843326694366626, -0.8334309368181174, -0.7919175817459623};
    move_joint_srv.request.is_joint_pose = 1;
    move_joint_srv.request.common.vel = 0.2;
    move_joint_srv.request.common.acc = 1.0;

    // call move joint service
    if (move_joint_client.call(move_joint_srv))
    {
        ROS_INFO("Successed to run move joint\n");
    }
    else
    {
        ROS_ERROR("Failed to run move joint");
    }


    //generate a move line srv data.
    lebai_msgs::MoveLine move_line_srv;
    move_line_srv.request.joint_pose = {-1.9979141024218048, -1.1726324385393299, 0.9936360553529241, -1.0942076707586763, -0.849345987492431, -0.8172282647460754};
    move_line_srv.request.is_joint_pose = 1;
    move_line_srv.request.common.vel = 0.2;
    move_line_srv.request.common.acc = 1.0;

    // call move joint service
    if (move_line_client.call(move_line_srv))
    {
        ROS_INFO("Successed to run move line\n");
    }
    else
    {
        ROS_ERROR("Failed to run move line");
    }
    return 0;
}